File: data_structures_py/LinkedList.py
from sqlalchemy import null


class Node:
	def __init__(self, value, next):
		self.value = value
		self.next = next

class Linkedlist:

	def __init__(self):
		self.head = Node(0, null)
	
	def insert(self, value, index=-1):
		if index < 0:
			# appending node at last
			current_node = self.head
			while True:
				if current_node.next == null:
					# last element here
					new_node = Node(value, null)
					current_node.next = new_node
					break
				current_node = current_node.next  # iterating to next node
		else:
			# appending node at index
			current_node = self.head
			for _ in range(index):
				current_node = current_node.next
			# ith ELEMENT HERE
			new_node = Node(value, current_node.next)
			current_node.next = new_node

	def delete(self, index=-1):
		if index < 0:
			# deleting last node
			current_node = self.head
			while True:
				if current_node.next.next == null:
					# last element here
					current_node.next = null
					break
				current_node = current_node.next  # iterating to next node
		else:
			# deleting node at index
			current_node = self.head
			for _ in range(index):
				current_node = current_node.next
			# ith ELEMENT HERE
			current_node.next = current_node.next.next

	def search(self, value):
		i = 0
		current_node = self.head.next
		while current_node != null:
			if current_node.value == value:
				return i
			current_node = current_node.next
			i += 1
		return False

	def printList(self):
		current_node = self.head
		while current_node.next != null:
			print(current_node.next.value)
			current_node = current_node.next



if __name__ == "__main__":
	ll = Linkedlist()
	ll.insert(23)
	ll.insert(75)
	ll.insert(10, 2)

	ll.delete(2)
	ll.search(75)

	ll.printList()

File: data_structures_cpp/LinkedList.cpp
#include <iostream>
using namespace std;

template <class T>
struct Node {
	T value;
	struct Node* next;
};
template <class T>
class LinkedList {
	public:
		struct Node<T>* head;
		LinkedList() {
			head = (struct Node<T>*) malloc(sizeof(struct Node<T>*));
			head->next = NULL;
		}
		insert(T value) {
			struct Node<T>* new_node = (struct Node<T>*) malloc(sizeof(struct Node<T>*));
			new_node -> value = value;
			new_node -> next = head -> next;
			head -> next = new_node;
		}
		remove () {
			if (head->next != NULL) {
				head -> next = head -> next -> next;
			}
		}
		sort () {
			struct Node<T>* temp;
			struct Node<T>* inner_temp;
			T swapVal;

			temp = head->next;
			while (temp != NULL) {
				inner_temp = temp->next;
				while (inner_temp != NULL) {
					if (temp->value > inner_temp->value) {
						// swapping values
						swapVal = temp->value;
						temp->value = inner_temp->value;
						inner_temp->value = swapVal;
					}
					inner_temp = inner_temp->next;
				}
				temp = temp->next;
			}
		}
		printList() {
			struct Node<T>* temp = head -> next;
			while (temp != NULL) {
				cout << temp -> value << endl;
				temp = temp -> next;
			}
		}
};

int main() {

	LinkedList<int> ll;

	ll.insert(34);
	ll.insert(56);
	ll.insert(12);

	ll.remove();

	ll.sort();

	ll.printList();

	return 0;
}
